File: public/dsa/searching/next_greater_element_in_sorted_array/right_binary_search/next_greater_element_in_sorted_array.cpp
#include <bits/stdc++.h>
using namespace std;

int right_binary_search(const vector<int> &nums, int target) {
    int left = 0, right = nums.size() - 1;
    while (left < right) {
        int mid = (left + right + 1) / 2;  // equivalent to ceil((left + right) / 2.0)
        if (nums[mid] > target) {
            right = mid - 1;
        } else {
            left = mid;
        }
    }
    return left + 1;
}

int getNextGreaterElement(vector<int> &arr, int key) {
    int r = right_binary_search(arr, key);
    if (r == arr.size()) {
        return key;
    }
    return arr[r];
}

// tests-start
int main() {
    vector<int> arr = {1, 2, 3, 3, 4, 4, 8, 10};
    int key = 4;
    cout << getNextGreaterElement(arr, key) << endl;
    return 0;
}
// tests-end

/*output
8
*/

File: public/dsa/searching/next_greater_element_in_sorted_array/right_binary_search/next_greater_element_in_sorted_array.py
from math import ceil
from typing import List


def right_binary_search(nums: List[int], target: int) -> int:
    left = -1
    right = len(nums) - 1
    while left < right:
        mid = ceil((left + right) / 2)
        if nums[mid] > target:
            right = mid - 1
        else:
            left = mid
    return left + 1


class Solution:
    def getNextGreaterElement(self, arr: List[int], key: int) -> int:
        r = right_binary_search(arr, key)
        if r == len(arr):
            return key
        return arr[r]


# tests-start
arr = [1, 2, 3, 3, 4, 4, 8, 10]
key = 4
sol = Solution()
print(sol.getNextGreaterElement(arr, key))
# tests-end

'''output
8
'''
